discovery/syncer.go

```diff
@@ -933,102 +964,99 @@ func (g *GossipSyncer) replyChanRangeQuery(query *lnwire.QueryChannelRange) erro
+		// the first in the range. We add one as even if all channels
 		// returned are in the same block, we need to count that.
-		numBlocksInResp := lastBlockHeight - firstBlockHeight + 1
+		numBlocks := lastHeight - firstHeight + 1
+		complete := uint8(0)
+		if finalChunk {
+			complete = 1
+		}
 
-		// With our chunk assembled, we'll now send to the remote peer
-		// the current chunk.
-		replyChunk := lnwire.ReplyChannelRange{
+		return g.cfg.sendToPeerSync(&lnwire.ReplyChannelRange{
 			QueryChannelRange: lnwire.QueryChannelRange{
 				ChainHash:        query.ChainHash,
-				NumBlocks:        numBlocksInResp,
-				FirstBlockHeight: firstBlockHeight,
+				NumBlocks:        numBlocks,
+				FirstBlockHeight: firstHeight,
 			},
-			Complete:     0,
+			Complete:     complete,
 			EncodingType: g.cfg.encodingType,
 			ShortChanIDs: channelChunk,
+		})
+	}
+
+	var (
+		firstHeight  = query.FirstBlockHeight
+		lastHeight   uint32
+		channelChunk []lnwire.ShortChannelID
+	)
+	for _, channelRange := range channelRanges {
+		channels := channelRange.Channels
+		numChannels := int32(len(channels))
+		numLeftToAdd := g.cfg.chunkSize - int32(len(channelChunk))
+
+		// Include the current block in the ongoing chunk if it can fit
+		// and move on to the next block.
+		if numChannels <= numLeftToAdd {
+			channelChunk = append(channelChunk, channels...)
+			continue
 		}
-		if isFinalChunk {
-			replyChunk.Complete = 1
-		}
-		if err := g.cfg.sendToPeerSync(&replyChunk); err != nil {
+
+		// Otherwise, we need to send our existing channel chunk as is
+		// as its own reply and start a new one for the current block.
+		// We'll mark the end of our current chunk as the height before
+		// the current block to ensure the whole query range is replied
+		// to.
+		log.Infof("GossipSyncer(%x): sending range chunk of size=%v",
+			g.cfg.peerPub[:], len(channelChunk))
+		lastHeight = channelRange.Height - 1
+		err := sendReplyForChunk(
+			channelChunk, firstHeight, lastHeight, false,
+		)
+		if err != nil {
 			return err
 		}
 
-		// If this was the final chunk, then we'll exit now as our
-		// response is now complete.
-		if isFinalChunk {
-			return nil
+		// With the reply constructed, we'll start tallying channels for
+		// our next one keeping in mind our chunk size. This may result
+		// in channels for this block being left out from the reply, but
+		// this isn't an issue since we'll randomly shuffle them and we
+		// assume a historical gossip sync is performed at a later time.
+		firstHeight = channelRange.Height
+		needSort := false
+		chunkSize := numChannels
+		if chunkSize > g.cfg.chunkSize {
+			rand.Shuffle(len(channels), func(i, j int) {
+				channels[i], channels[j] = channels[j], channels[i]
+			})
+			needSort = true
+			chunkSize = g.cfg.chunkSize
+		}
+		channelChunk = channels[:chunkSize]
+
+		// Sort the chunk once again if we had to shuffle it.
+		if needSort {
+			sort.Slice(channels, func(i, j int) bool {
```

Comment:
👍 nit: we only need to sort the `channelChunk` here, which will be faster if `numChannels >> chunkSize`